components/eppp_link/examples/emac2emac/main/app_main.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Unlicense OR CC0-1.0
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include "esp_system.h"
+#include "nvs_flash.h"
+#include "esp_event.h"
+#include "esp_netif.h"
+#include "eppp_link.h"
+#include "esp_log.h"
+#include "esp_check.h"
+#include "console_ping.h"
+#include "esp_eth_driver.h"
+#include "driver/gpio.h"
+#include "esp_eth_phy_dummy.h"
+
+#ifdef CONFIG_EXAMPLE_NODE_SERVER
+#define EPPP_CONFIG() EPPP_DEFAULT_SERVER_CONFIG()
+#define EPPP_ROLE EPPP_SERVER
+#define EPPP_RMII_CLK_SINK
+#else
+#define EPPP_CONFIG() EPPP_DEFAULT_CLIENT_CONFIG()
+#define EPPP_ROLE EPPP_CLIENT
+#define EPPP_RMII_CLK_SOURCE
+#endif
+
+void register_iperf(void);
+
+static const char *TAG = "eppp_emac2emac";
+
+
+static esp_eth_mac_t *s_mac = NULL;
+static esp_eth_phy_t *s_phy = NULL;
+
+#ifdef EPPP_RMII_CLK_SOURCE
+IRAM_ATTR static void gpio_isr_handler(void *arg)
+{
+    BaseType_t high_task_wakeup = pdFALSE;
+    TaskHandle_t task_handle = (TaskHandle_t)arg;
+
+    vTaskNotifyGiveFromISR(task_handle, &high_task_wakeup);
+    if (high_task_wakeup != pdFALSE) {
+        portYIELD_FROM_ISR();
+    }
+}
+#else
+#define STARTUP_DELAY_MS 500
+#endif
+
+esp_err_t eppp_transport_ethernet_init(esp_eth_handle_t *handle[])
+{
+    *handle = malloc(sizeof(esp_eth_handle_t));
+    ESP_RETURN_ON_FALSE(handle, ESP_ERR_NO_MEM, TAG, "Our of memory");
+
+#ifdef EPPP_RMII_CLK_SOURCE
+    esp_rom_gpio_pad_select_gpio(EMAC_CLK_OUT_180_GPIO);
+    gpio_set_pull_mode(EMAC_CLK_OUT_180_GPIO, GPIO_FLOATING); // to not affect GPIO0 (so the Sink Device could be flashed)
+    gpio_install_isr_service(0);
+    gpio_config_t gpio_source_cfg = {
+        .pin_bit_mask = (1ULL << CONFIG_EXAMPLE_RMII_CLK_READY_GPIO),
+        .mode = GPIO_MODE_INPUT,
+        .pull_up_en = GPIO_PULLUP_DISABLE,
+        .pull_down_en = GPIO_PULLDOWN_ENABLE,
+        .intr_type = GPIO_INTR_ANYEDGE
+    };
+    gpio_config(&gpio_source_cfg);
+    TaskHandle_t task_handle = xTaskGetHandle(pcTaskGetName(NULL));
+    gpio_isr_handler_add(CONFIG_EXAMPLE_RMII_CLK_READY_GPIO, gpio_isr_handler, task_handle);
+    ESP_LOGW(TAG, "waiting for RMII CLK sink device interrupt");
+    ESP_LOGW(TAG, "if RMII CLK sink device is already running, reset it by `EN` button");
```

Comment:
Should we make it part of the source device to automatically reset the device if the user enables it?